paddle/framework/tensor.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <typeinfo>
+#include "paddle/framework/ddim.h"
+#include "paddle/framework/enforce.h"
+#include "paddle/platform/assert.h"
+#include "paddle/platform/place.h"
+
+namespace paddle {
+namespace framework {
+
+class Tensor {
+  using paddle::platform::Place;
+  using paddle::platform::get_place;
+
+ public:
+  explicit Tensor(DDim dims) : dims_(dims), place_(get_place()) {}
+  explicit Tensor(DDim dims, Place place) : dims_(dims), place_(place) {}
+
+  Tensor& operator=(const Tensor& src) = delete;
+
+  template <typename T>
+  const T* data() const {
+    PADDLE_ENFORCE(holder_ != nullptr);
+    PADDLE_ENFORCE(holder_->Place() == place_);
+    PADDLE_ENFORCE(holder_->Size() >= Numel() * sizeof(T));
+    return static_cast<const T*>(holder->Ptr());
+  }
+
+  template <typename T>
+  bool NeedReset() const {
+    return (holder_ == nullptr || holder_->Place() != place_ ||
+            holder_->Size() < Numel() * sizeof(T));
+  }
+
+  // must be POD types
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* mutable_data() {
+    if (NeedReset<T>()) {
+      holder_.reset(new PlaceholderImpl(place_, Numel() * sizeof(T)));
+    }
+    return static_cast<T*>(holder_->Ptr());
+  }
+
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* mutable_data(const DDim& dims) {
+    dims_ = dims;
+    return mutable_data<T>();
+  }
+
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* mutable_data(const DDim& dims, const Place& place) {
+    dims_ = dims;
+    place_ = place;
+    return mutable_data<T>();
+  }
+
+  int Rank() const { return arity(dims_); }
+
+  int Numel() const { return product(dims_); }
+
+  void Reshape(const DDim& dims) {
```

Comment:
There is no need for Reshape, because we can simply call `mutable_data(new_place, new_dim)`.